### PR TITLE
Add batching and broadcasting support in shader compiler

### PR DESCRIPTION
**Pull request by dsmilkov**
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/219)

### CODE REVIEW

**Comment by nsthorat** [comment]
:lgtm_strong:

Wow. Great change. I have only cosmetic comments!

---

Reviewed 14 of 19 files at r1, 19 of 19 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/math/math.ts, line 783 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJCKnFhdRVboRhRF5:-KwgJCKnFhdRVboRhRF6:b-7gkyap) ([raw file](https://github.com/pair-code/deeplearnjs/blob/28772021dd093cc5fc10fcf970673e80c09f80eb/src/math/math.ts#L783)):*
> ```XML
>   /**
>    * Computes the softmax normalized vector given the logits.
>    * @param logits The logits vector.
> ```

no longer a vector

---

*[src/math/math.ts, line 800 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJffpo4B2DxQe1rI4:-KwgJffpo4B2DxQe1rI5:b-q4j0ar) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/math.ts#L800)):*
> ```XML
>         // Do it in log space for numerical stability.
>         // exp(X - logSumExp(X))
>         const lse = this.logSumExp(logits, [dim], true /* keepDim */);
> ```

s/keepDim/keepDims

---

*[src/math/math.ts, line 724 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-Kwjx_2aGAgVPMBZbTlY:-Kwjx_2bVUIIZV4Jwl4Z:bchif8z) ([raw file](https://github.com/pair-code/deeplearnjs/blob/950dc61ad8705b34004303ede86d89a8e044b7c2/src/math/math.ts#L724)):*
> ```XML
> 
>   /**
>    * Computes the minimum value from the input.
> ```

to be consistent with max, add the block talking about axis / keepdims

---

*[src/math/math_cpu.ts, line 538 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJxlQJjddJL1YRKpQ:-KwgJxlQJjddJL1YRKpR:b-rdt6hx) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/math_cpu.ts#L538)):*
> ```XML
>     if (!axis_util.axesAreInnerMostDims(axes, input.rank)) {
>       throw new Error(
>           `logSumExp reduction is supported only for the ` +
> ```

optionally add an axis_util.assertAxisAreInnerMostDims (with string for op) so we dont repeat ourselves

---

*[src/math/reduction_ops_test.ts, line 44 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-Kwjy5PJq5H8Bd20QfDB:-Kwjy5PJq5H8Bd20QfDC:b4ky3qv) ([raw file](https://github.com/pair-code/deeplearnjs/blob/950dc61ad8705b34004303ede86d89a8e044b7c2/src/math/reduction_ops_test.ts#L44)):*
> ```XML
>     it('2D', math => {
>       const a = Array2D.new([2, 3], [3, -1, 0, 100, -7, 2]);
>       expect(math.min(a).get()).toBeCloseTo(-7);
> ```

use test_util.expectNumbersClose, here and everywhere else in this file

---

*[src/math/transpose_test.ts, line 23 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgKiQGfHc-3d7xzF8E:-KwgKiQGfHc-3d7xzF8F:b-u4whw0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/transpose_test.ts#L23)):*
> ```XML
> import {Array2D, Array3D} from './ndarray';
> 
> // math.switchDim
> ```

update this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-:-Kwjymv3U4f2aRL_yUoG:b-luvqwo)*

**Comment by dsmilkov** [comment]
Review status: 23 of 28 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[src/math/math.ts, line 783 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJCKnFhdRVboRhRF5:-KwkE9JGgsSNtj_Na1Yl:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/28772021dd093cc5fc10fcf970673e80c09f80eb/src/math/math.ts#L783)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

no longer a vector
</blockquote></details>

Done.

---

*[src/math/math.ts, line 800 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJffpo4B2DxQe1rI4:-Kwk5Cxmx-ySVC_GPXwF:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/math.ts#L800)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/keepDim/keepDims
</blockquote></details>

Done.

---

*[src/math/math.ts, line 724 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-Kwjx_2aGAgVPMBZbTlY:-Kwk56reJ87aW3iKFCC8:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/950dc61ad8705b34004303ede86d89a8e044b7c2/src/math/math.ts#L724)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

to be consistent with max, add the block talking about axis / keepdims
</blockquote></details>

Done.

---

*[src/math/math_cpu.ts, line 538 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgJxlQJjddJL1YRKpQ:-KwkBsP8p90Y9RFxeSXD:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/math_cpu.ts#L538)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

optionally add an axis_util.assertAxisAreInnerMostDims (with string for op) so we dont repeat ourselves
</blockquote></details>

Done.

---

*[src/math/reduction_ops_test.ts, line 44 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-Kwjy5PJq5H8Bd20QfDB:-KwkDGCrT8iSsLi8Zh0y:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/950dc61ad8705b34004303ede86d89a8e044b7c2/src/math/reduction_ops_test.ts#L44)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use test_util.expectNumbersClose, here and everywhere else in this file
</blockquote></details>

Done.

---

*[src/math/transpose_test.ts, line 23 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219#-KwgKiQGfHc-3d7xzF8E:-KwkDJeQ7e7X3L62BFOH:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/592a255bdc7714c7d5fac63ea7390548a314978d/src/math/transpose_test.ts#L23)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

update this
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/219)*
